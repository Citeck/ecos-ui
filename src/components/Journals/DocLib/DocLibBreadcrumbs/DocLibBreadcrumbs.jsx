import React from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import { JournalUrlParams } from '../../../../constants';
import DocLibConverter from '../../../../dto/docLib';

import './DocLibBreadcrumbs.scss';

const DocLibBreadcrumbs = ({ path = [], onClick }) => {
  if (!Array.isArray(path) || !path.length) {
    return null;
  }

  const _onClick = id => typeof onClick === 'function' && onClick(id);

  const _renderItem = item => {
    const completeItem = DocLibConverter.completeItemId(item);
    const folderId = completeItem.id;
    const title = completeItem.disp;

    const urlParts = queryString.parseUrl(window.location.href);
    const query = urlParts.query;
    query[JournalUrlParams.DOCLIB_FOLDER_ID] = folderId;
    const urlQuery = queryString.stringify(query);
    const url = `${urlParts.url}?${urlQuery}`;

    return (
      <React.Fragment key={folderId}>
        <span className="ecos-doclib__breadcrumbs-link-separator">/</span>
        <a
          href={url}
          onClick={e => {
            e.preventDefault();
            _onClick(folderId);
          }}
          className="ecos-doclib__breadcrumbs-link"
        >
          {title}
        </a>
      </React.Fragment>
    );
  };

  return <div className="ecos-doclib__breadcrumbs">{path.map(_renderItem)}</div>;
};

DocLibBreadcrumbs.propTypes = {
  stateId: PropTypes.string,
  path: PropTypes.array,
  onClick: PropTypes.func
};

export default DocLibBreadcrumbs;
