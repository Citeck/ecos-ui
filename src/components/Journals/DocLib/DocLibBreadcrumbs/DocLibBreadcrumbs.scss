@import '../../../../styles/constants';

.ecos-doclib {
  &__breadcrumbs {
    margin-bottom: 1.5em;
    background-color: $color-white;
    border-radius: $border-radius-extra-large;
    padding: 15px 20px;

    box-sizing: border-box;
    direction: rtl;
    display: inline-block;
    max-width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;

    &-link,
    &-link-separator {
      font-size: $font-size-standard;
      font-weight: $font-weight-semi-bold;
      color: $text-color-gray;
    }

    &-link-separator {
      margin: 0 5px;
    }
  }
}
