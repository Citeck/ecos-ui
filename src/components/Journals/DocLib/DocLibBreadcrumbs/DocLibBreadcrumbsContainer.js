import { connect } from 'react-redux';

import { wrapArgs } from '../../../../helpers/redux';
import { openFolder } from '../../../../actions/docLib';
import { selectDocLibFolderPath } from '../../../../selectors/docLib';

import DocLibBreadcrumbs from './DocLibBreadcrumbs';

const mapStateToProps = (state, { stateId }) => {
  const path = selectDocLibFolderPath(state, stateId);

  return {
    path
  };
};

const mapDispatchToProps = (dispatch, props) => {
  const w = wrapArgs(props.stateId);
  return {
    onClick: id => {
      dispatch(openFolder(w(id)));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DocLibBreadcrumbs);
