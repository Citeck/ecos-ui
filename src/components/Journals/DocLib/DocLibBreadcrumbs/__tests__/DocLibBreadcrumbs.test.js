import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import DocLibBreadcrumbs from '../DocLibBreadcrumbs';

const demoItems = [
  {
    id: 'item1',
    disp: 'Folder 1',
  },
  {
    id: 'item2',
    disp: 'Folder 2',
  },
];

describe('DocLibBreadcrumbs tests', () => {
  describe('<DocLibBreadcrumbs />', () => {
    it('should render DocLibBreadcrumbs component', () => {
      const { container } = render(<DocLibBreadcrumbs />);
      expect(container).toMatchSnapshot();
    });

    it('should render all items', () => {
      const { container } = render(<DocLibBreadcrumbs path={demoItems} />);
      expect(container.getElementsByClassName('ecos-doclib__breadcrumbs-link')).toHaveLength(demoItems.length);
      expect(container.getElementsByClassName('ecos-doclib__breadcrumbs-link-separator')).toHaveLength(demoItems.length);
    });

    it('should render item label', () => {
      const { container, asFragment } = render(<DocLibBreadcrumbs path={demoItems} />);
      const link = container.getElementsByClassName('ecos-doclib__breadcrumbs-link');
      expect(link[0].textContent).toBe(demoItems[0].disp);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render null if empty path', () => {
      const { container, asFragment } = render(<DocLibBreadcrumbs path={[]} />);
      expect(container.outerHTML).toBe('<div></div>');
      expect(asFragment()).toMatchSnapshot();
    });

    it(`should call 'onClick' when click by item`, async () => {
      const user = userEvent.setup();

      const onClick = jest.fn();
      const elementIndex = 0;
      const { container } = render(<DocLibBreadcrumbs path={demoItems} onClick={onClick} />);

      await user.click(container.getElementsByClassName('ecos-doclib__breadcrumbs-link')[0]);

      expect(onClick).toHaveBeenCalledTimes(1);
      expect(onClick).toHaveBeenCalledWith(demoItems[elementIndex].id);
    });
  });
});
