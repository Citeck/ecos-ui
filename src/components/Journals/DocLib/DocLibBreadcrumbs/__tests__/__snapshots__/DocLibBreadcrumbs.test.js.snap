// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocLibBreadcrumbs tests <DocLibBreadcrumbs /> should render DocLibBreadcrumbs component 1`] = `<div />`;

exports[`DocLibBreadcrumbs tests <DocLibBreadcrumbs /> should render item label 1`] = `
<DocumentFragment>
  <div
    class="ecos-doclib__breadcrumbs"
  >
    <span
      class="ecos-doclib__breadcrumbs-link-separator"
    >
      /
    </span>
    <a
      class="ecos-doclib__breadcrumbs-link"
      href="http://localhost/?folderId=item1"
    >
      Folder 1
    </a>
    <span
      class="ecos-doclib__breadcrumbs-link-separator"
    >
      /
    </span>
    <a
      class="ecos-doclib__breadcrumbs-link"
      href="http://localhost/?folderId=item2"
    >
      Folder 2
    </a>
  </div>
</DocumentFragment>
`;

exports[`DocLibBreadcrumbs tests <DocLibBreadcrumbs /> should render null if empty path 1`] = `<DocumentFragment />`;
